New guide on migrating to module workers (#2898)
cloudflare/cloudflare-docs: products/workers/src/content/learning/migrating-to-module-workers.md

1. Durable Objects require the module syntax to work.
1. Module Workers do not rely on any global bindings, which means the Workers runtime does not need to set up fresh execution contexts, making Module Workers safer and faster to run.
1. Module Workers are ES Modules, which allows them to be shared and published to npm, for example. Module Workers can be imported by and composed within other Module Workers.

## Migrating a simple Worker

The following example demonstrates a Worker that redirects all incoming requests to a URL with a `301` status code.

With the Service Worker syntax, the example Worker looks like:

```javascript
async function handler(request) {
  const base = "https://example.com";
  const statusCode = 301;

  const destination = new URL(request.url, base);
  return Response.redirect(destination.toString(), statusCode);
}

// Initialize Worker
addEventListener("fetch", (event) => {
  event.respondWith(handler(event.request));
});
```

Module Workers replace the `addEventListener` syntax with an object definition, which must be the file's default export (via `export default`). The example code above becomes:

```javascript

> export default = {

Reviewer: ```suggestion
export default {
```

  fetch(request) {
    const base = "https://example.com";
    const statusCode = 301;

    const destination = new URL(request.url, base);
    return Response.redirect(destination.toString(), statusCode);
  },
};
```

## Accessing event or context data

> Workers often need access to data not in the `request` object. For example, sometimes Workers use [`waitUntil`](https://developers.cloudflare.com/workers/runtime-apis/fetch-event#waituntil) to delay execution. Module workers can access `waitUntil` vida the `context` parameter. For a list of Module worker parameters, see [this list](https://developers.cloudflare.com/workers/runtime-apis/fetch-event#parameters).

Reviewer: ```suggestion
Workers often need access to data not in the `request` object. For example, sometimes Workers use [`waitUntil`](/runtime-apis/fetch-event#waituntil) to delay execution. Module workers can access `waitUntil` vida the `context` parameter. For a list of Module worker parameters, see [this list](/runtime-apis/fetch-event#parameters).
```

This example code:

```javascript

> async function handler(event) {

Reviewer: This should align with `triggerEvent` or vice versa

---
order: 15
pcx-content-type: concept
---

# Migrating to module Workers

> This guide will show you how to migrate your Workers from the [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) syntax to the new [JavaScript module](https://blog.cloudflare.com/workers-javascript-modules/) format.

Reviewer: Optional, just to position the "Service Worker" and "Module Worker" names side by side.

```suggestion
This guide will show you how to migrate your Workers from the [Service Worker](https://developer.mozilla.org/en-US/docs/Web/API/Service_Worker_API) syntax to the new [Module Worker](https://blog.cloudflare.com/workers-javascript-modules/) format.
```

## Advantages of migrating

> There are several reasons you might want to migrate your Workers to the module Worker syntax:

Reviewer: ```suggestion
There are several reasons you might want to migrate your Workers to the module syntax:
```

> 1. Cloudflare Durable Objects require the module syntax to work.

Reviewer: ```suggestion
1. Durable Objects require the module syntax to work.
```

> 1. Module Workers do not require any global bindings to be set up behind the scenes, making them safer and faster to run.

Reviewer: ```suggestion
1. Module Workers do not rely on any global bindings, which means the Workers runtime does not need to set up fresh execution contexts, making Module Workers safer and faster to run.
```

> 1. Workers written in the new module format can be published to npm and shared; they can also be imported into other module Workers.

Reviewer: ```suggestion
1. Module Workers are ES Modules, which allows them to be shared and published to npm, for example. Module Workers can be imported by and composed within other Module Workers.
```

## Migrating a simple Worker

The following example demonstrates a Worker that redirects all incoming requests to a URL with a `301` status code.

With the Service Worker syntax, the example Worker looks like:

```javascript
async function handler(request) {

>   const base = "https://example.com";
>   const statusCode = 301;
> 
>   const url = new URL(request.url);
>   const { pathname, search } = url;
> 
>   const destinationURL = base + pathname + search;
>   return Response.redirect(destinationURL, statusCode);

Reviewer: ```suggestion
  const base = "https://example.com";
  const statusCode = 301;

  const destination = new URL(request.url, base);
  return Response.redirect(destination.toString(), statusCode);
```

}

// Initialize Worker
addEventListener("fetch", (event) => {
  event.respondWith(handler(event.request));
});
```

> JavaScript module Workers replace the `addEventListener` syntax with an object declared with `export default`. The example code above becomes:

Reviewer: ```suggestion
Module Workers replace the `addEventListener` syntax with an object definition, which must be the file's default export (via `export default`). The example code above becomes:
```

```javascript

> export default = {

Reviewer: ```suggestion
const worker = {
```

Author: Thanks! I had a comment on the gdoc to move to default but forgot I left the export named! I'll update!

  fetch(request) {

>     const base = 'https://example.com';
>     const statusCode = 301;
>     const url = new URL(request.url);
>     const { pathname, search } = url;
>     const destinationURL = base + pathname + search;
>     return Response.redirect(destinationURL, statusCode);

Reviewer: ```suggestion
    const base = "https://example.com";
    const statusCode = 301;
    
    const destination = new URL(request.url, base);
    return Response.redirect(destination.toString(), statusCode);
```

  },
};

> export default worker;

Reviewer: I changed the above to `const worker` to accommodate this line. Not sure which version you wanted

```

## Accessing event or context data

> Workers often need access to data not in the `request` object. For example, sometimes Workers use [`waitUntil`](https://developers.cloudflare.com/workers/runtime-apis/fetch-event#waituntil) to delay execution. Just like with the Service Worker syntax, module Workers have access to the [`fetchEvent` object](https://developers.cloudflare.com/workers/runtime-apis/fetch-event#supported-fetchevent-properties).

Reviewer: ```suggestion
Workers often need access to data not in the `request` object. For example, sometimes Workers use [`waitUntil`](https://developers.cloudflare.com/workers/runtime-apis/fetch-event#waituntil) to delay execution. Just like with the Service Worker syntax, Module Workers have access to the [`fetchEvent` object](https://developers.cloudflare.com/workers/runtime-apis/fetch-event#supported-fetchevent-properties).
```

Reviewer: But they don't have access to FetchEvent – it's exploded across `req` and `context` arguments.

Also the example is about scheduled event

> This means that:

Reviewer 2: ```suggestion
This example code:
```

```javascript
async function handler(event) {
  // Fetch some data
  console.log("cron processed", event.scheduledTime);
}

// Initialize Worker
addEventListener("scheduled", (event) => {
  event.waitUntil(handler(event));
});
```

> Becomes:

Reviewer 2: ```suggestion
Then becomes:
```

```javascript
async function triggerEvent(event) {
  // Fetch some data
  console.log("cron processed", event.scheduledTime);
}

const worker = {
  async scheduled(event, env, ctx) {
    ctx.waitUntil(triggerEvent(event));
  },
};

export default worker;
```

## Module Workers in the dashboard

The module Worker syntax has full support in the Cloudflare Dashboard. You can go to https://workers.new/, pick the HTTP Handler example and click “Create Service”. After it’s created, click “Quick Edit” and it will launch the dashboard. Click the “Send” button at the top to test if your service is returning “Hello world”.

> Now, you can replace:

Reviewer 2: ```suggestion
Then replace:
```

```javascript
addEventListener("fetch", (event) => {
  event.respondWith(handleRequest(event.request));
});

async function handleRequest(request) {
  return new Response("Hello world");
}
```

With:

```javascript
export default {
  fetch() {
    return new Response("Hello world");
  },
};
```

> Click “Save and Deploy” and then click “Send” again to test that your Worker is still returning “Hello world”.

Reviewer 2: ```suggestion
Select **Save and Deploy** and then select **Send** again to test that your Worker is still returning `“Hello world”`.
```

## Configuring your wrangler.toml

> To add support for module Workers to an existing project, you can update your `wrangler.toml` file like so:

Reviewer 2: ```suggestion
To add support for module Workers to an existing project, update your `wrangler.toml` file like so:
```
